### ISSUE TITLE
PARC Hyperparameters

### ISSUE THREAD

**Issue by jack89roberts** [alan-turing-institute/ARC-LoCoMoSeT#20]
Previously we were using (also used in the PARC author's repo):

```python
PCA(
     n_components=f,
     svd_solver='randomized',
     random_state=1919,
     iterated_power=1
)
```

then recently tried changing to just defaults:

```python
PCA(n_components=f)
```

This will either use a non-randomized solver or a randomized solver with more iterations than the original one (see https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html). It can take a lot longer as the number of images scales.

Understand the consequences of using one or the other (e.g. what does it mean for the PCA to not be fully optimised and does that matter for PARC?)

**Comment by eddableheath** [comment]
In the PARC paper they investigate the changes in mean Pearon's correlation across the datasets and models for each metric with different values of f, using the optimal value of f for each metric moving forward, where they suggest that f=32 is the optimal case for PARC. I would have thought that this relies on PCA being undertaken by the method they implement.

Futher to this I suggest we drop experimenting with different values of f in favour of sticking with f=32 as a hyperparameter.

**Comment by eddableheath** [comment]
If set to default the `svd_solver` parameter will enable 'randomised' anyway if f < 80% of the smallest dimension of the features, which is going to almost always be the case given the size of the models. However I think for uniformity this is kept as standard by the original authors. The full SVD would be ideal, however hugely inefficient hence the randomisation, and better to standardise across instances.

The `iterated_power` parameter controls the approximation depth of the randomised SVD algorithm, again this is for standardisation reasons as far as I can tell as otherwise this would vary dependent on the size of the input. 

I propose that we stick with the original values (with `f=32` as this is the value found to be optimal for PARC) as this subroutine is designed to standardise across the different model so keeping the PCA standard for all of them should ensure uniformity in the characterstics we wish to control for.

**Comment by jack89roberts** [comment]
Sure, that sounds like a good plan for WP1, but I think we should keep this issue around as something that could be explored further in later WPs.

I don't think they comment on the solver parameters in the paper/codebase anywhere? I'm still a bit uneasy about the consequences of having a (potentially) poorly fitted PCA but not familiar enough with the details of both PCA and PARC to say how that could impact the results, if at all.

**Comment by eddableheath** [comment]
I think that it's likely going to be fairly reasonable, the order 1 randomised SVD approximation algorithm it uses is relatively accurate, which sklearn will default to if the features dimension is > 40 anyway. Not sure why they've ensured this happens in all cases but this appears to me to be what would happen in most cases anyway given the default PCA parameters, for the sake of efficiency. (I've not see any models so far that we've been considering that have anywhere near as low feature dimensions).

**Comment by jack89roberts** [comment]
[The docs](https://scikit-learn.org/stable/modules/generated/sklearn.utils.extmath.randomized_svd.html#rf38c2b656ebc-1) give the default `iterated_power` value (which is passed to `num_iter` in the solver) as 4 or 7 fyi:

> **n_iter** int or ‘auto’, default=’auto’
Number of power iterations. It can be used to deal with very noisy problems. When ‘auto’, it is set to 4, unless n_components is small (< .1 * min(X.shape)) in which case n_iter is set to 7. This improves precision with few components. Note that in general users should rather increase n_oversamples before increasing n_iter as the principle of the randomized method is to avoid usage of these more costly power iterations steps. When n_components is equal or greater to the effective matrix rank and the spectrum does not present a slow decay, n_iter=0 or 1 should even work fine in theory (see [[1]](https://scikit-learn.org/stable/modules/generated/sklearn.utils.extmath.randomized_svd.html#rf38c2b656ebc-1) page 9).

But as before, happy to start with what was in the PARC repo and keep this as something to bear in mind later.

**Comment by eddableheath** [comment]
Are we leaving this issue open past WP1 as we want to experiment with these parameters later?

**Comment by jack89roberts** [comment]
Yeah I've removed it from the list of things for WP1 but let's leave it open (even if we end up not experimenting with it it's good to have this around as a reminder).

**Comment by jack89roberts** [comment]
Things to try:

- [x] PCA with default parameters (see discussion above)
- [x] Varying the feature reduction
  - [x] different values of f
  - [x] no feature reduction at all
- [x] Sanity check on a different dataset
- [x] Sanity check on more models 
- [x] Double-check implementation against author repo

**Comment by jack89roberts** [comment]
After trying most the above I haven't been able to improve the PARC results. Here's an example plot with more models than we used in WP1:

![actual_vs_parc](https://github.com/alan-turing-institute/ARC-LoCoMoSeT/assets/16308271/e3506458-d2e6-4380-8bdd-ef8bf95742fb)

I still plan to double-check the implementation, and we can keep PARC in the background as it's quite fast to run, but I don't think we can rely on it at this point.

**Comment by eddableheath** [comment]
Just had a look and there is a difference in the implementation that I somehow missed, in their implementation they use the `sklearn.preprocessing.StandardScaler()` method to Z-normalise the features prior to the computation of the correlation matrices (but post feature reduction).

**Comment by jack89roberts** [comment]
Ah nice spot, there's a chance that could change things (maybe trying scaling prior to PCA as well).

**Comment by jack89roberts** [comment]
I've started a run of PARC with scaling despite this, but adding the scaling causes `test_parc_perfect_features` to fail (I get a PARC score of about 82 instead of 100).

I don't think scaling the features should stop them from being "perfect features" (it's just subtracting and then dividing by a constant), so it's bizarre. Can investigate on Thursday.

**Comment by jack89roberts** [comment]
and it seems to get worse with increasing number of classes/features.

I did quickly check the scaler output and it was giving mean zero, standard deviation one for each feature as expected.

**Comment by eddableheath** [comment]
Closed by #52 PR.
